Add new spec for course approval

simfeld opened the pull request:
Hier ein Vorschlag für das Vorgehen in Bezug auf Page objects / Application actions anhand eines neuen Tests. In den js-files unter `pages` werden verschiedene Funktionen definiert, die in mehreren Tests wiederverwendet werden können. Einige à la Page objects (greifen auf das UI zu) und andere à la Application actions (lösen Requests aus).

Ich habe zudem in `course_edit.spec.js` versucht, einen Test zu schreiben, der die Äquivalenz von UI-Interaktionen und POST-Requests aufzeigen soll. Die JSON-API gibt jedoch nicht alle Daten des Events aus, daher ist der Check nicht vollständig. Eventuell finden wir hier noch eine bessere Lösung.

Für mich wäre folgendes Vorgehen denkbar (im Sinne von Application actions):
- UI-Interaktionen für eine bestimmte Funktionen werden einmal in einem Test geschrieben.
- Für weitere Tests wird diese Funktion mit Requests abgebildet und ausgelagert zur Wiederverwendung.

Merci für euer Feedback.

simfeld committed:
Add new spec for course approval

carlobeltrame commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,67 @@
+      const EVENT_ID = '84'
+      const GROUP_ID = '2'
+
+      editRequest(EVENT_ID, GROUP_ID, {
```
Könnte man den Payload noch in `editRequest` verarbeiten, sodass man von aussen etwas einfacheres hineingeben kann? Etwa so würde ich mir eine Verwendung von `editRequest` vorstellen:
```js
editRequest(EVENT_ID, GROUP_ID, {
  state: 'application_open',
  application_opening_at: '01.01.2020',
  application_closing_at: '31.12.2020',
  // ...
}
```
Am liebsten hätte ich es, wenn ich nur die wirklich geänderten Felder angeben muss, und `editRequest` den Rest mit Standardwerten ergänzt. Aber das ist vielleicht noch einen Schritt zu weit entfernt.

cleverer replied:
Wie meinst du Standardwerte? Fixe default werte oder dynamisch geladene?

Für beides kann man das mit dem [Spread operator](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Operators/Spread_operator) sehr einfach umsetzen…

carlobeltrame commented on .docker/cypress/spec/cypress/pages/person.js:
```diff
@@ -0,0 +1,19 @@
+export const imitate = (personID, groupID) => {
```
Ich fände es als Anwender dieser Funktion schöner, wenn man die Gruppen-Id nicht angeben müsste. Dazu könntest du zuerst ein `cy.visit` mit `log: false` auf die Person (`/de/people/${personID}`) machen, das leitet dich automatisch auf die URL inklusive Gruppe weiter. Danach könntest du `cy.request` mit `url: cy.url() + '/impersonate'` oder sogar mit `cy.get('Imitieren').click()` die Imitation starten. Nachteil ist natürlich der zusätzliche Netzwerk-Request für die User-Seite. Aber ich sehe Cypress als Tool zum automatisieren von Test-Abläufen die man sonst von Hand durcklicken muss, darum finde ich das jetzt nicht so schlimm.

cleverer replied:
Wenn möglich würde ich in Hilfsfunktionen/Commands auf `cy.visit()`verzichten (ist glaube ich auch best practice), da das die Tests extrem verlangsamt. Er wartet dann bis alle Resourcen geladen haben. Beim `cy.request()` ist das nicht der Fall.
Ich stimme dir aber zu, dass man das suchen der `groupID` automatisieren sollte 😉

carlobeltrame commented on .docker/cypress/spec/cypress/integration/course_edit.spec.js:
```diff
@@ -0,0 +1,57 @@
+    cy.get('textarea#event_location').focus().clear().type(FIELDS['event[location]'])
+    cy.get('select#event_state').select('Offen zur Anmeldung').should('have.value','application_open')
+    cy.get('input#event_motto').focus().clear().type(FIELDS['event[motto]'])
+    cy.get('ul.nav-tabs a[href="#application"]').click()
```
Der Lesbarkeit halber und gemäss [Cypress-Best-Practices](https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements) würde ich es vorziehen wenn wir wo immer möglich den Text (oder fontawesome-Icon-Namen) statt kryptischen CSS selectors zum selektieren verwenden.

Als Ausnahme wird auf der verlinkten Seite genannt, wenn der Text sich ändern darf und der Test weiterhin durchkommen soll. Die PBS kontrolliert aber sowohl diese E2E Tests wie auch die Translations zumindest des PBS Wagons. Ausserdem sollten in unserem Seed-Setup mit Transactions auch alle Personen und Gruppen immer gleich heissen. Daher denke ich dass sich solche Texte in unserem Fall nur sehr selten ändern werden.

Wo es nicht möglich ist, mit reiner Text-Selektion zum richtigen Element zu finden, finde ich CSS aber immer noch legitim. Aber ich überlege mir beim schreiben von E2E-Tests gern, wie ich den Ablauf jemandem übers Telefon erklären würde. Meist komme ich durch indem ich sage "Klicke auf 'Anmeldung', dann auf das Papierkorb-Icon".

cleverer replied:
Hier könnte ich mir auch diverse cypress-commands vorstellen (z.B. `cy.getHeaderTabs()`, `cy.getContent`). Da fände ich css-selektoren dann auch legitim, da wir die dann zentral an einer Stelle anpassen müssten…

carlobeltrame commented on .docker/cypress/spec/cypress/integration/course_edit.spec.js:
```diff
@@ -0,0 +1,57 @@
+    cy.get('select#event_state').select('Offen zur Anmeldung').should('have.value','application_open')
+    cy.get('input#event_motto').focus().clear().type(FIELDS['event[motto]'])
+    cy.get('ul.nav-tabs a[href="#application"]').click()
+    cy.get('input#event_application_opening_at').focus().clear().type(FIELDS['event[application_opening_at]'])
```
Für Inputs wirds etwas schwieriger, mit dem Text zu arbeiten. Eine "korrekte" aber etwas komplizierte Lösung wäre aus meiner Sicht:
```js
cy.get('Anmeldebeginn').closest('.control-group').within(() => {
  return cy.get('input:not([type='hidden']), textarea').first()
})
```
Könnte man so etwas vielleicht in eine eigene selector-Funktion `cy.input` oder so ähnlich packen? Kenne mich noch zu wenig mit Cypress aus um da konkrete Vorschläge zu machen. Am liebsten würde ich es dann so verwenden: `cy.input('Anmeldebeginn').focus().clear().type(...)`

cleverer replied:
Ich würde es wohl eher so lösen:

```javascript
cy.get('Anmeldebeginn').closest('.control-group').find('input:not([type='hidden']), textarea').first()
```

Wenn möglich sollten wir callbacks vermeiden (zwecks [retry-ability](https://docs.cypress.io/guides/core-concepts/retry-ability.html#Commands-vs-assertions))…

carlobeltrame replied:
Ah okay, muss mich in dem Fall mal informieren was der Unterschied zwischen `.within` und `.find` ist.

carlobeltrame commented on .docker/cypress/spec/cypress/integration/course_edit.spec.js:
```diff
@@ -0,0 +1,57 @@
+import { editRequest } from '../pages/event.js'
+
+describe('A course', function () {
+  const EVENT_ID = '113'
```
Bietet Cypress Möglichkeiten für uns damit wir mehr mit Bezeichnungen und weniger mit Nummern arbeiten müssen? Es wäre aus meiner Sicht schon geholfen wenn wir diese IDs global benennen könnten (z.B. TOP_LAYER_LEADER oder irgend so etwas, ähnlich wie in hitobito selber die Fixtures funktionieren (`people(:top_layer_leader)`)? Ich fände es viel lesbarer wenn überall Namen statt IDs stehen und würde dafür auch das wiederholte Tippen des Namen in Kauf nehmen.

cleverer replied:
Soweit ich weiss, gibts sowas nicht in cypress. Braucht es in meinen Augen, aber auch nicht, einfach ein JS-File im support-ordner reicht doch für solche konstanten?

cleverer reviewed (commented):
Ich habe mir nochmals die [Best Practices](https://docs.cypress.io/api/cypress-api/custom-commands.html#Best-Practices) zu Cypress [Custom Commands](https://docs.cypress.io/api/cypress-api/custom-commands.html#Syntax) durchgelesen. Ich schlage vor, dass wir uns so weit wie möglich daran halten.
Dasselbe mit der [Ordner-Struktur](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Folder-Structure).

Aufgrund der Best-Practices denke ich, wir sollten aus `imitate()` und `editRequest()` Cypress Commands machen (und entsprechend umbenennen) und aus `register()` und `approve()` lokale Funktionen (wie im Beispiel [hier](https://docs.cypress.io/api/cypress-api/custom-commands.html#Best-Practices)).
In den Cypress-Commands sollten wir wenn möglich allen Output unterdrücken und nur eine log Message ausgeben: [Command Logging](https://docs.cypress.io/api/cypress-api/custom-commands.html#Command-Logging). Sonst werden die Test-Runs schnell unübersichtlich. Das würde auch gegen assertion in den Commands sprechen (wenn würde ich da normale-Javascript checks bevorzugen welche dann eine Exception werfen).

Bezgl. Test Namen: da fände ich es toll, wenn wir konsistent "implizite Subjekte" gemäss [Writing Assertions](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Writing-Assertions), [Assertions](https://example.cypress.io/commands/assertions#assertions) und [Behaviour Driven Development](https://mochajs.org/#bdd) haben könnten .

simfeld committed:
Improve test style

- Manipulate payload in editRequest
- New command getFullUrl which finds the group id from a given event or
  or person id
- Query elements based on text where possible
- Shared file for constants

simfeld commented:
Ich habe einige Anpassungen basierend auf euren Reviews vorgenommen, vorerst mal bei `course_edit.spec.js`.

Cypress Custom Commands: Ich bin der Meinung, dass nur global verwendbare Funktionen als Cypress Custom Command geschrieben werden sollten (im Sinne von [Don’t make everything a custom command](https://docs.cypress.io/api/cypress-api/custom-commands.html#1-Don%E2%80%99t-make-everything-a-custom-command)). Dazu zählt für mich z. B. `login` oder `getCSRFToken`. Domain-spezifische Funtionen wie `editRequest` für events sollten wir meiner Ansicht nach in separaten Files auslagern und bei Bedarf importieren. Da bin ich auch dafür, soweit wie möglich auf das UI zu verzichten und mit Requests zu arbeiten (solange irgendwo die gleiche Funktionalität analog via UI getestet wird). Vielleicht ist `pages` aber nicht der beste übergeordnete Name für diese Files :)

Test-Namen: @cleverer Kannst du etwas genauer ausführen, worauf du mit "implizite Subjekte" hinauswillst? Meine Erfahrungen mit BDD sind noch relativ beschränkt.

cleverer commented on .docker/cypress/spec/cypress/pages/event.js:
```diff
@@ -5,15 +5,21 @@ export const register = (eventID, groupID) => {
+export const editRequest = (eventID, payload) => {
+  cy.getFullUrl('event', eventID).then(res => {
+    cy.getCSRFToken().then(token => {
+      Object.keys(payload).forEach(key => {
+        payload[`event[${key}]`] = payload[key]
+        delete payload[key]
+      })
```
Mach doch hier ein neues Objekt, dann must du die alten keys nicht löschen. (Wegen [Immutable_object](https://en.wikipedia.org/wiki/Immutable_object) und so…)

cleverer commented on .docker/cypress/spec/cypress/support/commands.js:
```diff
@@ -65,6 +65,42 @@ Cypress.Commands.add("getCSRFToken", () => {
   })
 })
 
+Cypress.Commands.add('getInputByText', { prevSubject: 'optional'}, (subject, text, checkbox) => {
```
Finde den Command so relativ verwirrend bzw. verstehe nicht auf den ersten Blick, was er macht.
Mein erster Vorschlag: zwei Commands, mit sprechenderen Namen: `nearestCheckbox` und `nearestInput`.
Ein Indiz für komplexen Code, den man in zwei Funktionen packen sollte/könnte ist jeweils ein if/else…

Zweiter Vorschlag (das wäre in jedem fall sinnvoll): Schreib doch einen kurzen Doc-Kommentar, der den Command beschreibt…

simfeld replied:
Du hast absolut recht, der ganze Command ist überhaupt keine Augenweide. Ich mache zwei draus und dokumentiere noch.

cleverer commented on .docker/cypress/spec/cypress/support/commands.js:
```diff
@@ -65,6 +65,42 @@ Cypress.Commands.add("getCSRFToken", () => {
 
+Cypress.Commands.add('getInputByText', { prevSubject: 'optional'}, (subject, text, checkbox) => {
+  if (subject) {
+    if (checkbox)
+      return cy.wrap(subject, { log: false }).contains(text).closest('.control-group', { log: false })
+        .contains(checkbox).find('input[type="checkbox"]', { log: false }).first({ log: false })
+      return cy.wrap(subject, { log: false }).contains(text).closest('.control-group', { log: false })
+        .find('input:not([type="hidden"]), textarea, select', { log: false }).first({ log: false })
```
Ich bin grundsätzlich fan von kurzen Schreibweisen, allerdings leidet in meinen Augen hier die Lesbarkeit. Ich versuche grundsätzlich if/else immer mit geschweiften Klammern zu schreiben. Als Ausnahme dazu sehe ich kurze Statements, welche dann den Scope verlassen (z.B. `return false`). Hier (und auch weiter unten) finde ich aber die beiden returns sehr verwirrend…

cleverer commented on .docker/cypress/spec/cypress/support/commands.js:
```diff
@@ -65,6 +65,42 @@ Cypress.Commands.add("getCSRFToken", () => {
+})
+
+Cypress.Commands.add('getFullUrl', (type, id) => {
+  if (type == 'person') {
```
Hier wäre wohl ein switch besser als ein ife/elseif/else 😉

Noch einfacher fände ich aber auch hier zwei Commands. Das könnte man dann aber so Refactoren, dass "Gegen Aussen" zwei Commands existieren, die aber dann eine interne Funktion mit jeweils anderen Parametern (die URL und das Regex) aufrufen.

cleverer commented on .docker/cypress/spec/cypress/support/constants.js:
```diff
@@ -0,0 +1,22 @@
+export const root_user = {
+  login: 'hitobito-pbs@puzzle.ch',
```
Detail: für mich besteht ein Login aus Username und Passwort, was dann so aussehen würde:

```javascript
export const root_login = {
   username: 'hitobito-pbs@puzzle.ch',
   password: 'hito42bito'
}
```

Ich versuche jeweils auch Abkürzungen in Variablen zu vermeiden, dank autocompletion muss man heutzutage ja nicht mehr tippen 😉

cleverer commented on .docker/cypress/spec/cypress/support/constants.js:
```diff
@@ -0,0 +1,22 @@
+  pw: 'hito42bito'
+}
+
+export const unspunne_smw = {
```
Mir ist leider nicht ganz klar was `unspunne_smw` sein soll… vor allem wenn der Name `unspunne`nicht bekannt ist 😝

simfeld replied:
So heisst das Lager im UI. Unspunne Sicherheitsmodul Wasser. Mir ist leider nichts schlaueres eingefallen, bin aber offen für Vorschläge :-)

cleverer replied:
Fände `unspunne_sicherheitsmodul_wasser` schon viel klarer 😉

(`unspunne_simo_wasser ` wäre auch ok, aber wenn wir schon klar sind, können wirs auch gleich ganz durchziehen)

cleverer reviewed (commented):
(Habe mir jetzt nur die neuen Changes angeschaut…)

> Cypress Custom Commands: Ich bin der Meinung, dass nur global verwendbare Funktionen als Cypress Custom Command geschrieben werden sollten (im Sinne von Don’t make everything a custom command). Dazu zählt für mich z. B. login oder getCSRFToken. Domain-spezifische Funtionen wie editRequest für events sollten wir meiner Ansicht nach in separaten Files auslagern und bei Bedarf importieren. Da bin ich auch dafür, soweit wie möglich auf das UI zu verzichten und mit Requests zu arbeiten (solange irgendwo die gleiche Funktionalität analog via UI getestet wird). Vielleicht ist pages aber nicht der beste übergeordnete Name für diese Files :)

Absolut einverstanden, passt für mich so.

Bezgl. BDD/Testnamen: `course_edit.spec.js` finde ich super so. Im `course_approval.spec.js` hats noch Potential. Wenn ich da die `context` und `it` aneinandergereiht lese ergibt das keine schönen Sätze:
- "Applications overview … Approval turns the warning badge into a success badge"
- "Applications overview … New participations show up under applications with warning badge, and success badge if approved"

Dazu im Vergleich:

* "A course … can be edited via the UI or a request, the result is the same".

simfeld committed:
Refactor Cypress commands, add request functions

cleverer committed:
Some Refactoring

cleverer committed:
Make course_register.spec.js work reliably…

cleverer commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,120 @@
+    cy.get('@event_url').then((url) => {
+      cy.visit(`${url}/application_market`)
+    })
+    cy.get(`tbody#participants tr:contains('${full_name}')`)
+      .find('td:eq(3) > span').should('have.attr','title', 'Kursfreigabe ausstehend')
```
Hier checkst du ja nur das title-attribut des badges, nicht aber das eigentliche icon… Ist das gewollt so?
```suggestion
    cy.get(`tbody#participants tr:contains('${full_name}')`)
      .find('td:eq(3) > span').should('contain', '?').should('have.attr','title', 'Kursfreigabe ausstehend')
```

cleverer commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,120 @@
+    cy.get('@event_url').then((url) => {
+      cy.visit(`${url}/application_market`)
+    })
+    cy.get(`tbody#participants tr:contains('${full_name}')`)
+      .find('td:eq(3) > span').should('have.attr','title', 'Kursfreigabe bestätigt')
```
```suggestion
    cy.get(`tbody#participants tr:contains('${full_name}')`)
      .find('td:eq(3) > span').should('contain', '✓').should('have.attr','title', 'Kursfreigabe bestätigt')
```

cleverer commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,120 @@
+    // check the badge
+    cy.get('@event_url').then((url) => {
+      cy.visit(`${url}/application_market`)
+      cy.get(`tbody#applications:contains('${full_name}')`)
+        .find('tr > td:eq(4) > span').should('have.attr','title', 'Kursfreigabe ausstehend')
```
Ich würde hier nur noch den Inhalt checken, das TItle attribut haben wir ja schon getestet.
```suggestion
      cy.get(`tbody#applications:contains('${full_name}')`)
        .find('tr > td:eq(4) > span').should('contain', '?')
```

cleverer commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,120 @@
+    // check the badge
+    cy.get('@event_url').then((url) => {
+      cy.visit(`${url}/application_market`)
+      cy.get(`tbody#applications:contains('${full_name}')`)
+        .find('tr > td:eq(4) > span').should('have.attr','title', 'Kursfreigabe bestätigt')
```
```suggestion
      cy.get(`tbody#applications:contains('${full_name}')`)
        .find('tr > td:eq(4) > span').should('contain', '✓')
```

cleverer commented on .docker/cypress/spec/cypress/integration/course_approval.spec.js:
```diff
@@ -0,0 +1,120 @@
+    })
+  })
+
+  it('are correctly propagated to the approval overview', function () {
+    approve(necessitatibus.event.id, necessitatibus.event.participation_id, necessitatibus.event.fields)
+    cy.getEventUrl(necessitatibus.event.id).then((res) => {
+      cy.visit(`${res.url}/approvals`)
+    })
+    cy.contains('Bewertung').parent().should('contain', necessitatibus.event.fields.occupation_assessment)
+  })
+})
```
Hier wäre es toll, wenn du vor dem approve noch den negativen-test durchführst, also dass die Freigabe noch nicht existiert.

Am besten checkst du auch gleich hier den badge inkl. inhalt

cleverer reviewed (commented):
Ich habe zwei kleine commits gemacht mit ein wenig refactoring…

Du testest die helper-funktionen per json-API/Datenbankabfrage. Ich denke es wäre hier zielführender, über das UI zu testen (ich weiss, mühsam), aber dafür wissen wir dann, dass das UI ebenfalls funktionert.
Das macht dann auch Hilfsfunktionen wie `replace_empty_strings_with_null`(siehe https://github.com/cleverer/hitobito-docker/pull/5/files#diff-9310007a61edaba18bb275428cc4ff10R4) nötig.

Bei den PageHelpers (Wie ich sie jetzt nenne…) wäre es sehr hilfreich, wenn du noch ein Cypress logging einbaust: https://github.com/cleverer/hitobito-docker/pull/5/files#diff-5d4ed4027ab8a04d87c3f7ccccab247cR97-R100 und dafür die requests und wraps etc. nicht loggst. Das macht die runs einiges lesbarer.
